docs/user-guide/ingress.md

```diff
@@ -0,0 +1,328 @@
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Ingress
+
+**Table of Contents**
+<!-- BEGIN MUNGE: GENERATED_TOC -->
+
+- [Ingress](#ingress)
+  - [What is Ingress?](#what-is-ingress)
+  - [The Ingress Resource](#the-ingress-resource)
+  - [Ingress controllers](#ingress-controllers)
+  - [Types of Ingress](#types-of-ingress)
+    - [Single Service Ingress](#single-service-ingress)
+    - [Simple fanout](#simple-fanout)
+    - [Name based virtual hosting](#name-based-virtual-hosting)
+    - [Loadbalancing](#loadbalancing)
+  - [Updating an Ingress](#updating-an-ingress)
+  - [Future Work](#future-work)
+  - [Alternatives](#alternatives)
+
+<!-- END MUNGE: GENERATED_TOC -->
+
+__Terminology__
+
+Throughout this doc you will see a few terms that are sometimes used interchangably elsewhere, that might cause confusion. This section attempts to clarify them.
+
+* Node: A single virtual or physical machine in a Kubernetes cluster.
+* Cluster: A group of nodes firewalled from the internet, that are the primary compute resources managed by Kubernetes.
+* Edge router: A router that enforces the firewall policy for your cluster. This could be a gateway managed by a cloudprovider or a physical piece of hardware.
+* Cluster network: A set of links, logical or physical, that facilitate communication within a cluster according to the [Kubernetes networking model](https://github.com/kubernetes/kubernetes/blob/release-1.0/docs/admin/networking.md). Examples of a Cluster network include Overlays such as [flannel](https://github.com/coreos/flannel#flannel) or SDNs such as [OVS](https://github.com/kubernetes/kubernetes/blob/release-1.0/docs/admin/ovs-networking.md).
+* Service: A Kubernetes [Service](https://github.com/kubernetes/kubernetes/blob/release-1.0/docs/user-guide/services.md) that identifies a set of pods using label selectors. Unless mentioned otherwise, Services are assumed to have virtual IPs only routable within the cluster network.
+
+## What is Ingress?
+
+Typically, services and pods have IPs only routable by the cluster network. All traffic that ends up at an edge router is either dropped or forwarded elsewhere. Conceptually, this might look like:
+
+```
+    internet
+        |
+  ------------
+  [ Services ]
+```
+
+An Ingress is a collection of rules that allow inbound connections to reach the cluster services.
+
+```
+    internet
+        |
+   [ Ingress ]
+   --|-----|--
+   [ Services ]
+```
+
+It can be configured to give services externally-reachable urls, load balance traffic, terminate SSL, offer name based virtual hosting etc. Users request ingress by POSTing the Ingress resource to the API server. An [Ingress controller](#ingress-controllers) is responsible for fulfilling the Ingress, usually with a loadbalancer, though it may also configure your edge router or additional frontends to help handle the traffic in an HA manner.
+
+## The Ingress Resource
+
+A minimal Ingress might look like:
+
+```yaml
+01. apiVersion: extensions/v1beta1
```

Comment:
I prefer showing a working config in its entirety and explaning the whole thing. IMO yaml comments work better when you want to draw attention to something, not explain how it works. I could do it in 2 phases where i show the whole thing, then chunk it up to explain, if you'd prefer that I'll do it when i send the follow up pr explaining the controller.